Orleans.Consensus/Log/StreamLog.cs

```diff
@@ -0,0 +1,191 @@
+﻿namespace Orleans.Consensus.Log
+{
+    using Orleans.Consensus.Contract.Log;
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
+
+    public class StreamLog<TOperation> : IPersistentLog<TOperation>, IDisposable
+    {
+        const int CACHE_SIZE = 1000;
+
+        Stream stream;
+        ISerializer<LogEntry<TOperation>> serializer;
+        BTree<long, long> bTreeIndex;
+        IDictionary<long, LogEntry<TOperation>> cache;
+
+        public StreamLog(Stream stream, ISerializer<LogEntry<TOperation>> serializer)
+        {
+            if (null == stream) throw new ArgumentNullException(nameof(stream));
+            if (null == serializer) throw new ArgumentNullException(nameof(serializer));
+
+            this.stream = stream;
+            this.serializer = serializer;
+            this.bTreeIndex = new BTree<long, long>();
+            this.cache = new Dictionary<long, LogEntry<TOperation>>(CACHE_SIZE);
+            BuildIndex();
+        }
+
+        void BuildIndex()
+        {
+            stream.Position = 0;
+            var first = true;
+            while (stream.Position < stream.Length)
+            {
+                var position = stream.Position;
+                var entry = serializer.Deserialize(stream).Result;
```

Comment:
I'm not aware of any serializers which work in async, maybe we should just make it synchronous. Marc Gravell suggests async serialization is not worthwhile here: http://stackoverflow.com/a/11187593/635314, and that serializing to a buffer may be better (maybe RecyclableMemoryStream if we need it)